include/unifex/indexed_for.hpp

```diff
@@ -0,0 +1,169 @@
+
+  friend constexpr auto tag_invoke(
+      tag_t<blocking>,
+      const indexed_for_sender& sender) {
+    return blocking(sender.pred_);
+  }
+
+  template <typename Receiver>
+  struct indexed_for_receiver {
+    UNIFEX_NO_UNIQUE_ADDRESS Func func_;
+    UNIFEX_NO_UNIQUE_ADDRESS Policy policy_;
+    UNIFEX_NO_UNIQUE_ADDRESS Range range_;
+    UNIFEX_NO_UNIQUE_ADDRESS Receiver receiver_;
+
+    // sequenced_policy version supports forward range
+    template<typename... Values>
+    void apply_func_with_policy(const execution::sequenced_policy& policy, Range&& range, Func&& func, Values&... values)
+        noexcept(noexcept(std::invoke((Func &&) func_, std::declval<typename Range::iterator::value_type>(), values...))) {
+      for(auto idx : range) {
+        std::invoke((Func &&) func, idx, values...);
+      }
+    }
+
+    // parallel_policy version requires random access range
+    template<typename... Values>
+    void apply_func_with_policy(const execution::parallel_policy& policy, Range&& range, Func&& func, Values&... values)
+        noexcept(noexcept(std::invoke((Func &&) func_, std::declval<typename Range::iterator::value_type>(), values...))) {
+      auto start = range.begin();
+      for(auto idx = 0; idx < range.size(); ++idx) {
+        std::invoke((Func &&) func, start[idx], values...);
+      }
+    }
+
+    template <typename... Values>
+    void set_value(Values&&... values) && noexcept {
+      if constexpr (noexcept(std::invoke(
+                        (Func &&) func_, std::declval<typename Range::iterator::value_type>(), values...))) {
+        apply_func_with_policy(policy_, (Range&&) range_, (Func &&) func_, values...);
+        unifex::set_value((Receiver &&) receiver_, (Values &&) values...);
+      } else {
+        try {
+          apply_func_with_policy(policy_, (Range&&) range_, (Func &&) func_, values...);
+          unifex::set_value((Receiver &&) receiver_, (Values &&) values...);
+        } catch (...) {
+          unifex::set_error((Receiver &&) receiver_, std::current_exception());
+        }
+      }
+    }
+
+    template <typename Error>
+    void set_error(Error&& error) && noexcept {
+      unifex::set_error((Receiver &&) receiver_, (Error &&) error);
+    }
+
+    void set_done() && noexcept {
+      unifex::set_done((Receiver &&) receiver_);
+    }
+
+    template <
+        typename CPO,
+        std::enable_if_t<!is_receiver_cpo_v<CPO>, int> = 0>
+    friend auto tag_invoke(CPO cpo, const indexed_for_receiver& r) noexcept(
+        std::is_nothrow_invocable_v<CPO, const Receiver&>)
+        -> std::invoke_result_t<CPO, const Receiver&> {
+      return std::move(cpo)(std::as_const(r.receiver_));
+    }
+
+    template <typename Visit>
+    friend void tag_invoke(
+        tag_t<visit_continuations>,
+        const indexed_for_receiver& r,
+        Visit&& visit) {
+      std::invoke(visit, r.receiver_);
+    }
+  };
+
+  template <typename Receiver>
+  auto connect(Receiver&& receiver) && {
```

Comment:
This method should probably be constrained to only allow receivers that are able to receive the appropriate result types.